erpnext/accounts/doctype/coupon_code/coupon_code.js

```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2018, Frappe Technologies Pvt. Ltd. and contributors
+// For license information, please see license.txt
+
+frappe.ui.form.on('Coupon Code', {
+	coupon_name:function(frm){
+		if (frm.doc.__islocal===1) {
+			frm.trigger("make_coupon_code");
+		}
+	},
+	coupon_type:function(frm){
+		if (frm.doc.__islocal===1) {
+			frm.trigger("make_coupon_code");
+		}
+	},
+	make_coupon_code: function(frm) {
+		var coupon_name=frm.doc.coupon_name;
+		var coupon_code;
+		if (frm.doc.coupon_type=='Gift Card') {
+			coupon_code=Math.random().toString(12).substring(2, 12).toUpperCase();
+		}
+		else if(frm.doc.coupon_type=='Promotional'){
+			coupon_name=coupon_name.replace(/\s/g,'');
+			coupon_code=coupon_name.toUpperCase().slice(0,8);
+		}
+		frm.doc.coupon_code=coupon_code;
+		frm.refresh_field('coupon_code');
```

Comment:
Instead of JS can you generate coupon at py side, so it can be generate using API too. Also why two fields coupon_name and coupon_code? name can be code?

Reply:
IMO, coupon_code is usually short and that makes it tough to remember who it was targeted at and when it was run.  For instance, did we use "HolidaySave10" last year or was that "SaveHoliday10"?  Having coupon_name allows us to add "2019 Holiday Coupon - General Online Use"

Reply:
@parsimonyERP what @rohitwaghchaure is saying does not affect this functionality but will make it friendly for those creating coupon codes via an API

Reply:
> Instead of JS can you generate coupon at py side, so it can be generate using API too. Also why two fields coupon_name and coupon_code? name can be code?

It was on .py side originally. [Link](https://github.com/frappe/erpnext/blob/101ae0518da5a697cf21b7d024e11d47d07e5f02/erpnext/accounts/doctype/coupon_code/coupon_code.py)
Based on rmehta's comment it was shifted to JS
coupon_code - is for external consumption. Generally it is 8 char without space. Ex DELL30
coupon_name - is for internal ERPNext

Reply:
But in that case how to generate the coupon code using API?

Reply:
@rohitwaghchaure , I have updated autoname to generate coupon_code in case when JS is not invoked i.e. via API. Thanks! for the comment